Ignore ::-webkit-scrollbar styles when scrollbar-color is not auto

lukewarlow opened the pull request:
#### f056156a42b6e3c32320293c36b6e0f44451f509
<pre>
Ignore ::-webkit-scrollbar styles when scrollbar-color is not auto
<a href="https://bugs.webkit.org/show_bug.cgi?id=257617">https://bugs.webkit.org/show_bug.cgi?id=257617</a>

Reviewed by Tim Nguyen.

Rename hasCustomScrollbarStyle to usesLegacyScrollbarStyle to avoid confusion.

RenderStyle::usesLegacyScrollbarStyle() now returns true if a custom style
for the scrollbar has been set via ::-webkit-scrollbar, the value of
scrollbar-width is auto (default), and the value of scrollbar-color is auto (default).

* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-color-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-color-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-color-004-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-color-005-expected.txt:

* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::rootElementForCustomScrollbarPartStyle const):
(WebCore::LocalFrameView::createScrollbar):
(WebCore::LocalFrameView::canShowNonOverlayScrollbars const):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::canUseOverlayScrollbars const):
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::createScrollbar):
* Source/WebCore/rendering/RenderListBox.cpp:
(WebCore::RenderListBox::createScrollbar):
* Source/WebCore/rendering/RenderMenuList.cpp:
(RenderMenuList::createScrollbar):
* Source/WebCore/rendering/RenderSearchField.cpp:
(WebCore::RenderSearchField::createScrollbar):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::usesLegacyScrollbarStyle const):
(WebCore::RenderStyle::hasCustomScrollbarStyle const): Deleted.

Canonical link: <a href="https://commits.webkit.org/265423@main">https://commits.webkit.org/265423@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/78c9078f22c9092129ffa3c6b00a5f8bfddeb496

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10843 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11060 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/11365 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12491 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10392 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/10859 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/13434 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11039 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13299 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/11002 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11905 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/9125 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12896 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9198 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9782 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/17037 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10269 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9934 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13188 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10408 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/8492 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9568 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2596 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13841 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10269 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/rendering/style/RenderStyleInlines.h:
```diff
@@ -723,7 +723,7 @@ constexpr LengthType RenderStyle::zeroLength() { return LengthType::Fixed; }
 
 // ignore non-standard ::-webkit-scrollbar when standard properties are in use
-inline bool RenderStyle::hasCustomScrollbarStyle() const { return hasPseudoStyle(PseudoId::Scrollbar) && scrollbarWidth() == ScrollbarWidth::Auto; }
+inline bool RenderStyle::hasCustomScrollbarStyle() const { return hasPseudoStyle(PseudoId::Scrollbar) && scrollbarWidth() == ScrollbarWidth::Auto && !scrollbarColor().has_value(); }
```
I now realize hasCustomScrollbarStyle is a confusing name, it's more `hasScrollbarPseudoStyle` or `hasLegacyScrollbarStyle`.

Can we rename?

lukewarlow replied:
Ah yeah I see what you mean. I guess it's technically correct on the basis that the Legacy styling creates a Custom Scrollbar as opposed to a Native Scrollbar but I agree it can be better. I'll go with `usesLegacyScrollbarStyle`

nt1m replied:
Actually, even the verb `has` is confusing here, we should use the verb `uses` too.

lukewarlow committed:
Ignore ::-webkit-scrollbar styles when scrollbar-color is not auto
https://bugs.webkit.org/show_bug.cgi?id=257617

Reviewed by Tim Nguyen.

Rename hasCustomScrollbarStyle to usesLegacyScrollbarStyle to avoid confusion.

RenderStyle::usesLegacyScrollbarStyle() now returns true if a custom style
for the scrollbar has been set via ::-webkit-scrollbar, the value of
scrollbar-width is auto (default), and the value of scrollbar-color is auto (default).

* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-color-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-color-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-color-004-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-scrollbars/scrollbar-color-005-expected.txt:

* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::rootElementForCustomScrollbarPartStyle const):
(WebCore::LocalFrameView::createScrollbar):
(WebCore::LocalFrameView::canShowNonOverlayScrollbars const):
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::canUseOverlayScrollbars const):
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::createScrollbar):
* Source/WebCore/rendering/RenderListBox.cpp:
(WebCore::RenderListBox::createScrollbar):
* Source/WebCore/rendering/RenderMenuList.cpp:
(RenderMenuList::createScrollbar):
* Source/WebCore/rendering/RenderSearchField.cpp:
(WebCore::RenderSearchField::createScrollbar):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::usesLegacyScrollbarStyle const):
(WebCore::RenderStyle::hasCustomScrollbarStyle const): Deleted.

Canonical link: https://commits.webkit.org/265423@main
